### PR TITLE
Speedup symmetric eigendecomposition

### PR DESCRIPTION
**Pull request by t4c1**
## Summary

This pull request adds CPU implementation of an efficient symmetric eigensolver.

## Tests

Tests check that errors made by the new symmetric eigensolver and its components tridiagonalization and MRRR are not too large.

## Side Effects

/

## Checklist

- [x] Math issue #1156

- [x] Copyright holder: Tadej Ciglarič, University of Ljubljana

- [x] the basic tests are passing

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by bgoodri** [comment]
This seems nice. Is there a reason why Eigen didn't want this algorithm upstream?

**Comment by rok-cesnovar** [comment]
Just to clarify, do you mean in a sense if @t4c1 tried to get his merged in Eigen or why hasnt Eigen tried to implement this? 

For the former, we have not tried to merge this to Eigen. Tadej, Erik and others that work on RHMC want to speedup eigensolvers in Stan/Stan Math so getting this merged to Eigen hasnt really been discussed so far.

For the latter, they had this in plans for Eigen 3.0 (http://eigen.tuxfamily.org/index.php?title=Todo_for_3.0), have some parts implemented (https://eigen.tuxfamily.org/dox/BlockHouseholder_8h_source.html) but no full implementation as of yet.

**Comment by bgoodri** [comment]
That's fine. I just wanted to verify that Eigen hadn't said something like
"this is too inaccurate to include".

**Comment by syclik** [comment]
Is there anyone that can review this? @rok-cesnovar?

I'll take a look to make sure that it has enough tests / doc to pass. Then we just need someone to look at it to make sure it's correct.

File: stan/math/prim/mat/fun/mrrr.hpp [hunk 1]
```diff
@@ -0,0 +1,709 @@
...
+  int level;
+};
+
+/**
+ * Calculates eigenvalues and eigenvectors of a irreducible tridiagonal matrix T
+ * using MRRR algorithm. Use `tridiagonal_eigensolver` if any subdiagonal
+ * element might be (very close to) zero.
+ * @param diagonal Diagonal of of T.
+ * @param subdiagonal Subdiagonal of T.
+ * @param[out] eigenvalues Eigenvlues.
```
Comment (syclik) [inline stan/math/prim/mat/fun/mrrr.hpp]: Spelling error. "Eigenvlues" -> "Eigenvalues"

File: stan/math/prim/mat/fun/mrrr.hpp [hunk 2]
```diff
@@ -0,0 +1,709 @@
...
+
+struct mrrr_task {
+  int start, end;
+  double shift;  // total shift, not just the last one
+  Eigen::VectorXd l, d;
+  int level;
+};
+
+/**
+ * Calculates eigenvalues and eigenvectors of a irreducible tridiagonal matrix T
```
Comment (syclik) [inline stan/math/prim/mat/fun/mrrr.hpp]: typo: "a irreducible" to "an irreducible"

File: stan/math/prim/mat/fun/mrrr.hpp [hunk 3]
```diff
@@ -0,0 +1,709 @@
...
+ * using MRRR algorithm. Use `tridiagonal_eigensolver` if any subdiagonal
```
Comment (syclik) [inline stan/math/prim/mat/fun/mrrr.hpp]: Can you add a link to the "MRRR algorithm" or at least spell it out once? (Meaning... how do I search and find the algorithm online... I don't know who MRRR is and it's not really getting me anything useful.)

**Comment by rok-cesnovar** [comment]
I am not the most familiar with the maths behind eigendecomposition but I can dive into that if no one else wants to :) This has been sitting here for too long.

File: stan/math/prim/mat/fun/mrrr.hpp [hunk 4]
```diff
@@ -0,0 +1,709 @@
...
+ * Calculates eigenvalues and eigenvectors of a tridiagonal matrix T using MRRR
+ * algorithm. If a subdiagonal element is close to zero compared to neighbors on
+ * diagonal the problem can be split into smaller ones.
+ * @param diagonal Diagonal of of T.
+ * @param subdiagonal Subdiagonal of T.
+ * @param[out] eigenvalues Eigenvlues.
+ * @param[out] eigenvectors Eigenvectors.
+ * @param split_threshold Threshold for splitting the problem
+ */
+void tridiagonal_eigensolver(const Eigen::VectorXd& diagonal,
```
Comment (syclik) [inline stan/math/prim/mat/fun/mrrr.hpp]: If this is the actual function that's supposed to be called, please call the name of this file `tridiagonal_eigensolver.hpp`.

**Comment by syclik** [comment]
Agreed that this is sitting too long.

I think what would have helped: @t4c1, can you write up the issue more clearly? More specifically, for the very first post of #1156, can you write out:

1. The function signature you want to add.
2. What the function is supposed to do?

Reading the text of the issue, it's not clear what's happening. And there are very sparse details here in this PR, but a lot of code. I'm trusting you to know what you're talking about and if we can validate that

1. the code quality is good (relative to our metrics: doc + testing)
2. and we're pretty sure it does what you say it does

then we should be able to merge it without everyone having to know every implementation detail.

File: stan/math/prim/mat/fun/mrrr.hpp [hunk 5]
```diff
@@ -0,0 +1,709 @@
...
+/**
+ * Calculates eigenvalues and eigenvectors of a tridiagonal matrix T using MRRR
+ * algorithm. If a subdiagonal element is close to zero compared to neighbors on
+ * diagonal the problem can be split into smaller ones.
+ * @param diagonal Diagonal of of T.
+ * @param subdiagonal Subdiagonal of T.
+ * @param[out] eigenvalues Eigenvlues.
+ * @param[out] eigenvectors Eigenvectors.
+ * @param split_threshold Threshold for splitting the problem
+ */
```
Comment (syclik) [inline stan/math/prim/mat/fun/mrrr.hpp]: Does this function throw?

File: stan/math/prim/mat/fun/mrrr.hpp [hunk 6]
```diff
@@ -0,0 +1,709 @@
...
+
+/**
+ * Calculates eigenvalues and eigenvectors of a tridiagonal matrix T using MRRR
+ * algorithm. If a subdiagonal element is close to zero compared to neighbors on
+ * diagonal the problem can be split into smaller ones.
+ * @param diagonal Diagonal of of T.
+ * @param subdiagonal Subdiagonal of T.
+ * @param[out] eigenvalues Eigenvlues.
+ * @param[out] eigenvectors Eigenvectors.
+ * @param split_threshold Threshold for splitting the problem
```
Comment (syclik) [inline stan/math/prim/mat/fun/mrrr.hpp]: can you doc the default value?

File: stan/math/prim/mat/fun/mrrr.hpp [hunk 7]
```diff
@@ -0,0 +1,709 @@
...
+ */
+void tridiagonal_eigensolver(const Eigen::VectorXd& diagonal,
+                             const Eigen::VectorXd& subdiagonal,
+                             Eigen::VectorXd& eigenvalues,
+                             Eigen::MatrixXd& eigenvectors,
+                             const double split_threshold = 1e-12) {
```
Comment (syclik) [inline stan/math/prim/mat/fun/mrrr.hpp]: why is this the default value?

**Comment by syclik** [comment]
It's fairly clean, but I do need guidance. I can see me looking at this for hours on end without help because there's so much code here.

**Comment by syclik** [comment]
Also, I can pick this up again and do the review for code quality (not looking at the algorithm implementation) once I'm told how to look at this PR.

**Comment by SteveBronder** [comment]
@t4c1 have yinz reached out to the eigen team? If some benchmarks for speed / precision are made + some doc info I think this would have a bigger impact if it goes upstream

**Comment by t4c1** [comment]
I guess this would make more sense in Eigen library. This was my first PR so at the time I did not know what goes where. I haven't contacted Eigen team but I will. I will close this for now and reopen if there are any complications merging this to eigen.

**Pull request closed without merging**